Mathlib/Tactic/CancelDenoms.lean

```diff
@@ -0,0 +1,268 @@
+    let (v1, t1) := findCancelFactor e1
+    let (v2, t2) := findCancelFactor e2
+    let lcm := v1.lcm v2
+    (lcm, .node lcm t1 t2)
+  | (`HMul.hMul, #[_, _, _, _, e1, e2]) =>
+    let (v1, t1) := findCancelFactor e1
+    let (v2, t2) := findCancelFactor e2
+    let pd := v1 * v2
+    (pd, .node pd t1 t2)
+  | (`HDiv.hDiv, #[_, _, _, _, e1, e2]) =>
+    match isRatLit e2 with
+    | some q =>
+      let (v1, t1) := findCancelFactor e1
+      let n := v1.lcm q.num.natAbs
+      (n, .node n t1 <| .node q.num.natAbs .nil .nil)
+    | none => (1, .node 1 .nil .nil)
+  | (`Neg.neg, #[_, _, e]) => findCancelFactor e
+  | _ => (1, .node 1 .nil .nil)
+
+def synthesizeUsingNormNum (type : Expr) : MetaM Expr := do
+  synthesizeUsingTactic' type (← `(tactic| norm_num))
+
+/--
+`mkProdPrf n tr e` produces a proof of `n*e = e'`, where numeric denominators have been
+canceled in `e'`, distributing `n` proportionally according to `tr`.
+-/
+partial def mkProdPrf (α : Q(Type u)) (sα : Q(Field $α)) (v : ℕ) (t : Tree ℕ)
+    (e : Q($α)) : MetaM Expr := do
+  let amwo ← synthInstanceQ q(AddMonoidWithOne $α)
+  match t, e with
+  | .node _ lhs rhs, ~q($e1 + $e2) => do
+    let v1 ← mkProdPrf α sα v lhs e1
+    let v2 ← mkProdPrf α sα v rhs e2
+    mkAppM `CancelFactors.add_subst #[v1, v2]
+  | .node _ lhs rhs, ~q($e1 - $e2) => do
+    let v1 ← mkProdPrf α sα v lhs e1
+    let v2 ← mkProdPrf α sα v rhs e2
+    mkAppM `CancelFactors.sub_subst #[v1, v2]
+  | .node _ lhs@(.node ln _ _) rhs, ~q($e1 * $e2) => do
+    let v1 ← mkProdPrf α sα ln lhs e1
+    let v2 ← mkProdPrf α sα (v / ln) rhs e2
+    have ln' := (← mkOfNat α amwo <| mkRawNatLit ln).1
+    have vln' := (← mkOfNat α amwo <| mkRawNatLit (v/ln)).1
+    have v' := (← mkOfNat α amwo <| mkRawNatLit v).1
+    let ntp : Q(Prop) := q($ln' * $vln' = $v')
+    let npf ← synthesizeUsingNormNum ntp
+    mkAppM `CancelFactors.mul_subst #[v1, v2, npf]
+  | .node n lhs (.node rn _ _), ~q($e1 / $e2) => do
+    let v1 ← mkProdPrf α sα (v / rn) lhs e1
+    have rn' := (← mkOfNat α amwo <| mkRawNatLit rn).1
+    have vrn' := (← mkOfNat α amwo <| mkRawNatLit <| v / rn).1
+    have n' := (← mkOfNat α amwo <| mkRawNatLit <| n).1
+    have v' := (← mkOfNat α amwo <| mkRawNatLit <| v).1
+    let ntp : Q(Prop) := q($rn' / $e2 = 1)
+    let npf ← synthesizeUsingNormNum ntp
+    let ntp2 : Q(Prop) := q($vrn' * $n' = $v')
+    let npf2 ← synthesizeUsingNormNum ntp2
+    mkAppM `CancelFactors.div_subst #[v1, npf, npf2]
+  | t, ~q(-$e) => do
+    let v ← mkProdPrf α sα v t e
+    mkAppM `CancelFactors.neg_subst #[v]
+  | _, _ => do
+    have v' := (← mkOfNat α amwo <| mkRawNatLit <| v).1
+    let e' ← mkAppM `HMul.hMul #[v', e]
+    mkEqRefl e'
+
+def inferTypeQ' (e : Expr) : MetaM ((u : Level) × (α : Q(Type $u)) × Q($α)) := do
+  let α ← inferType e
+  let .sort u ← whnf (← inferType α) | throwError "not a type{indentExpr α}"
+  let some u' := u.dec | throwError "is a Prop, not a Type{indentExpr e}"
+  pure ⟨u', α, e⟩
```

Comment:
The QQ library has `inferTypeQ`, which returns the type using `Sort u` rather than `Type u`, but things like `Field` want a `Type u`. I'm not sure if there's another way I should be doing it, or if this function should find its way into QQ.

Reply:
I've wanted this as well, so I'd hope it can eventually be upstreamed to `QQ`. @gebner?

In the meantime, leave a comment to the effect that it might be upstreamable. Perhaps even put it under `Mathlib/QQ/`??

Reply:
Lol, apparently you wanted it so much that you managed to retroactively add it: https://github.com/leanprover-community/mathlib4/pull/526/files#diff-673a39ed7826ee06bdbde9432e777e3e62a4a8b3e48bf6aacafe939ed346e0adR37

In the meantime it was moved to `Util/Qq.lean`.  I've deleted mine, and I just took this as an opportunity to improve the docstring and try to follow the implementation of the real `inferTypeQ` a little closer (and avoid `unreachable!`).

I didn't leave a comment about upstreaming, but I figure that anything in `Util/Qq.lean` is implicitly fair game.